app/controllers/users/sessions_controller.rb

```diff
@@ -1,13 +1,89 @@
 # frozen_string_literal: true
 
 class Users::SessionsController < Devise::SessionsController
+  before_action :redirect_to_dashboard, only: %i[sign_in_with_token redirect_from_magic_link]
+  before_action :ensure_login_token_valid, only: %i[sign_in_with_token redirect_from_magic_link]
+
   def create
-    self.resource = warden.authenticate!(auth_options)
+    self.resource = validate_email
     if resource.errors.any?
       render :new
     else
+      self.resource = warden.authenticate!(auth_options)
       sign_in(resource_name, resource)
       respond_with resource, location: after_sign_in_path_for(resource)
     end
+  rescue Devise::Strategies::PasswordlessAuthenticatable::Error
+    render :login_email_sent
+  end
+
+  def sign_in_with_token
+    @user.update!(login_token: nil, login_token_valid_until: 1.year.ago)
+    sign_in(@user, scope: :user)
+    redirect_to after_sign_in_path_for(@user)
+  end
+
+  def redirect_from_magic_link
+    @login_token = params[:login_token] if params[:login_token].present?
```

Comment:
Is it finding any existing login_token from the session cookies?

Reply:
Nah, it's finding it from the url. The url we generate will look something like 
```
https://ecf-engage-and-learn-review-pr-128.london.cloudapps.digital/users/confirm-sign-in?login_token=81ed78c85706afc0546b
```
Note the last part: `?login_token=81ed78c85706afc0546b`. `?` in it means that the main part of the url is over, and that we are starting to include url parameters. `login_token` is the name of the parameter, and `81ed78c85706afc0546b` is the value.

And Rails helpfully makes this available in the controller `params`.

Reply:
I guess if it was not in the url, but you edited your session cookie to include `login_token` key, it might be able to find it, maybe. Although probably not inside `params`.

Reply:
I think i was thinking about how a session persists when you leave the site and then come back but are still logged in. I thought this could be doing that as well as what you've mentioned above.  But yeah thinking about it, that wouldn't be happening here because it wouldn't enter from the redirect from magic link.